src/ott/solvers/linear/univariate.py

```diff
@@ -12,134 +12,247 @@
+      subsampling, taken into account marginal probabilities.
+    quantiles: When a vector or a number of quantiles is passed, the distance
+      is computed by evaluating the cost function on the sectional (one for each
+      dimension) quantiles of the two point cloud distributions described in the
+      problem.
   """
 
   def __init__(
       self,
-      sort_fn: Optional[Callable[[jnp.ndarray], jnp.ndarray]] = None,
-      cost_fn: Optional[costs.CostFn] = None,
-      method: Literal["subsample", "quantile", "wasserstein",
-                      "equal"] = "subsample",
-      n_subsamples: int = 100,
+      num_subsamples: Optional[int] = None,
+      quantiles: Optional[Union[int, jnp.ndarray]] = None,
   ):
-    self.sort_fn = jnp.sort if sort_fn is None else sort_fn
-    self.cost_fn = costs.PNormP(2) if cost_fn is None else cost_fn
-    self.method = method
-    self.n_subsamples = n_subsamples
+    self._quantiles = quantiles
+    self.num_subsamples = num_subsamples
+
+  @property
+  def quantiles(self):
+    """Quantiles' values used to evaluate OT cost."""
+    if self._quantiles is None:
+      return None
+    if isinstance(self._quantiles, int):
+      return jnp.linspace(0.0, 1.0, self._quantiles)
+    return self._quantiles
+
+  @property
+  def num_quantiles(self):
+    """Number of quantiles used to evaluate OT cost."""
+    return 0 if self._quantiles is None else self.quantiles.shape[0]
 
   def __call__(
       self,
-      x: jnp.ndarray,
-      y: jnp.ndarray,
-      a: Optional[jnp.ndarray] = None,
-      b: Optional[jnp.ndarray] = None
+      prob: linear_problem.LinearProblem,
+      rng: Optional[jax.Array] = None,
   ) -> float:
-    """Computes the Univariate OT Distance between `x` and `y`.
+    """Computes Univariate Distance between the `d` dimensional slices.
 
     Args:
-      x: The first distribution of shape ``[n,]`` or ``[n, 1]``.
-      y: The second distribution of shape ``[m,]`` or ``[m, 1]``.
-      a: The first marginals when ``method = 'wasserstein'``. If :obj:`None`,
-        uniform will be used.
-      b: The second marginals when ``method = 'wasserstein'``. If :obj:`None`,
-        uniform will be used.
+      prob: describing, in its geometry attribute, the two point clouds
+        ``x`` and ``y`` (of respective sizes ``[n,d]`` and ``[m,d]``) and
+        a ground ``cost_fn`` for between two scalars. The ``[n,]`` and ``[m,]``
+        size probability weights vectors are stored in attributes ``a`` and
+        ``b``.
+      rng: used for random downsampling, if used.
+      return_transport: whether to return an average transport matrix (across
+        slices). Not available when approximating the distance computation
+        using subsamples, or quantiles.
 
     Returns:
-      The OT distance.
+      The OT distance, and possibly the transport matrix averaged by
+        considering all matrices arising from 1D transport on each of the ``d``
```

Comment:
No need to dedent.

Reply:
thanks. I think this was wrong overall, and needed rewriting! thanks for catching